### PR TITLE
Add support for building Windows pause image

### PR DESCRIPTION
**Pull request by claudiubelu**
**What type of PR is this?**

/kind feature

/sig windows

**What this PR does / why we need it**:

We can use ``docker buildx`` in order to build and push Windows images from the same Linux node, as long as the ``Dockerfile`` does not have any ``RUN`` commands in the Windows step.

We also need to create a non-default builder instance in order to be able to build and push Windows images.

The Windows images have to be built and pushed directly to the registry. Because of this, the make target ``push`` has been removed (the target ``all`` will build and push the images).

We need wincat for a few kubectl proxy scenarios.

For Windows containers without Hyper-V isolation, the host OS Version and the Container OS Version need to match, which is why we added multiple Windows OS Versions to the building process.

Adds support for Windows OS Versions: 1809, 1903, 1909, 2004.

Bumped pause image version to 3.4.

Co-Authored-By: Claudiu Belu <cbelu@cloudbasesolutions.com>
Co-Authored-By: Ben Moss <bmoss@pivotal.io>

Signed-off-by: Leah Hanson <lhanson@pivotal.io>

**Which issue(s) this PR fixes**:

Fixes #

**Special notes for your reviewer**:

**Does this PR introduce a user-facing change?**:

```release-note
Support for Windows container images (OS Versions: 1809, 1903, 1909, 2004) was added the pause:3.4 image.
```

**Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.**:

```docs

```

### CODE REVIEW

File: build/pause/Makefile
Comment (dims) [inline build/pause/Makefile:75]: bad merge? :(

**Comment by dims** [comment]
cc @endocrimes

File: build/pause/Makefile
Comment (claudiubelu) [reply build/pause/Makefile]: Oups. :) I had to rebase this PR a few times. :)

File: build/pause/Makefile
Comment (claudiubelu) [reply build/pause/Makefile:29]: It is used on L130:

    -t $(IMAGE_WITH_OS_ARCH)-${OSVERSION}:$(TAG) --build-arg BASE=${BASE}:${OSVERSION} --build-arg ARCH=$(ARCH) -f Dockerfile.windows .

File: build/pause/Dockerfile.windows
Comment (jeremyje) [inline build/pause/Dockerfile.windows:1]: For the sake of syntax highlighting in IDEs I think it makes sense to put this in `build/pause/windows/Dockerfile` and move linux to a similar location, `build/pause/linux/Dockerfile`. What do you think?

We may have to alter the docker context on `docker build`.

File: build/pause/Dockerfile.windows
Comment (jeremyje) [inline build/pause/Dockerfile.windows:23]: Replace `C:\Windows` with `%SystemRoot%`.
Replace `C:\Program Files\` with `%ProgramFiles%`.

File: build/pause/Dockerfile
Comment (jeremyje) [inline build/pause/Dockerfile:18]: Just curious how complex would it be to simply do a multi-platform build with linux and windows?

**Review by jeremyje — CHANGES_REQUESTED** [review]

File: build/pause/windows/wincat/wincat.go
Comment (benmoss) [reply build/pause/windows/wincat/wincat.go:31]: We were aiming to have the same arguments as netcat, ie `nc 127.0.0.1 6443` / `wincat 127.0.0.1 6443`

File: build/pause/Makefile
Comment (benmoss) [inline build/pause/Makefile:101]: Why do we want `all-container-docker` to only build Linux? This seems confusing. Why not just always build all OSes?

**Review by benmoss — COMMENTED** [review]
Thanks for picking this up @claudiubelu!

File: build/pause/Makefile
Comment (claudiubelu) [reply build/pause/Makefile:101]: There's no point to it.

First of all, if you are making the target ``all-container-docker`` probably means you only want that image locally, and thus, it probably means it's a Linux node. There's no need to build a Windows image in this case.

Secondly, if you are on a Linux node, unless you push the built images to a registry (``docker buildx build --output=type=registry``), that Windows image is not even visible locally like the rest of the images (if you run ``docker images -a``)

So, I just decided that it's not worth building the Windows image if the image is not meant to be pushed to a registry. There's no point to it.

Comment (benmoss) [reply build/pause/Makefile:101]: Ah, I get that. I still think it may be confusing that `make all` is not really making "all", but now that I get that `buildx build --output=docker` can't output Windows images the `all-container-docker` not including Windows makes more sense.

I was looking to test that the changes work without pushing them to a registry, which afaik isn't possible right now.

File: build/pause/Dockerfile_windows
Comment (benmoss) [inline build/pause/Dockerfile_windows:19]:
```suggestion
ADD bin/wincat.exe /Windows/System32/wincat.exe
```

File: build/pause/Makefile
Comment (benmoss) [inline build/pause/Makefile:120]:
```suggestion
	CGO_ENABLED=0 GOOS=windows GOARCH=${ARCH} go build -o $@ windows/wincat/wincat.go
```

**Review by benmoss — CHANGES_REQUESTED** [review]
I would move the wincat.exe to be output into the `bin` directory like the other binaries. This directory is already gitignored.

File: build/pause/Makefile
Comment (benmoss) [inline build/pause/Makefile:120]:
```suggestion
bin/wincat-windows-${ARCH}: windows/wincat/wincat.go
	CGO_ENABLED=0 GOOS=windows GOARCH=${ARCH} go build -o $@ $^
```

File: build/pause/Dockerfile_windows
Comment (marosset) [inline build/pause/Dockerfile_windows:21]: Should we add a link to https://github.com/moby/buildkit/issues/1560 in the comment so we can remove the note if/when the the issue gets resolved?

File: build/pause/Makefile
Comment (marosset) [inline build/pause/Makefile:86]: Has an issue/feature request been raised for this `docker manifest annotate doesn't allow us to set the os.version`?
If so should we link it here? If not should we file one?

Comment (claudiubelu) [reply build/pause/Makefile:86]: Good point, I've reported it now: https://github.com/moby/moby/issues/41417

Comment (claudiubelu) [reply build/pause/Makefile:86]: Interesting. But from what I see, that commit is not yet included in any release. We should add a TODO to mention this.

But I think it is going to take a bit until it is available. First, it needs to be released, then there should be a new Google Cloud Builder image released that includes that change.

**Comment by benmoss** [comment]
LGTM, but since I was pretty involved in the original implementation I don't think it's appropriate for me to sign-off.

**Comment by marosset** [comment]
LGTM

File: build/pause/Makefile
Comment (BenTheElder) [inline build/pause/Makefile:97]: just checking: will this work correctly in cloudbuild?

**Comment by BenTheElder** [comment]
/sig node
/sig release
Is there some history on the windows sources? (a previous image somewhere, or someone from node / windows can comment on these binaries?)

File: build/pause/windows/pause.c
Comment (BenTheElder) [inline build/pause/windows/pause.c:1]: I'm curious about the mix of C and go here, is there some background on this image?
I feel these binaries could use a little documentation, I wouldn't block the PR on it, but I think that would be helpful in the future.

pause.c seems pretty self explanatory, but I'm not sure why wincat even exists 🙃

**Comment by BenTheElder** [comment]
thank you @claudiubelu for working on the buildx switch here, this seems nice.  I'm attempting to catch up on reviews I've gotten behind on, I know you have a few here in the set I'm behind on.

I only have some minor comments on my first pass here but should probably take a closer look at the shell bits later.
I see now:
> We need wincat for a few kubectl proxy scenarios.

Can we explain that a bit more and check it in somewhere?

**Comment by claudiubelu** [comment]
> 
> 
> /sig node
> /sig release
> Is there some history on the windows sources? (a previous image somewhere, or someone from node / windows can comment on these binaries?)

I assume you mean for ``wincat.go``, right? If so, the code was taken almost as is from here: https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/cmd/wincat , and the original author, Ben Moss, gave us his blessing.

wincat is being used for port-forwarding, and this is the current usage [1]. We've been including wincat into our Windows pause images for a while [2]. Because of this, the port-forwarding tests work.

[1] https://github.com/kubernetes/kubernetes/blob/master/pkg/kubelet/dockershim/docker_streaming_windows.go#L31
[2] https://github.com/kubernetes-sigs/sig-windows-tools/blob/master/cmd/wincat/Dockerfile

File: build/pause/windows/pause.c
Comment (claudiubelu) [reply build/pause/windows/pause.c:1]: Hm, Linux has a ``pause.c`` too, so we're not doing anything new here either. https://github.com/kubernetes/kubernetes/blob/master/build/pause/pause.c

As for wincat, I've explained it on this PR's general comment section.

File: build/pause/Makefile
Comment (claudiubelu) [reply build/pause/Makefile:97]: I've built this image inside of a ``gcr.io/k8s-testimages/gcb-docker-gcloud:v20200422-b25d964`` container, which is the image mentioned in the pause image's ``cloudbuild.yaml`` file. It built succesfully, including the Windows images. See logs below.

Logs: https://paste.ubuntu.com/p/3ZjF4nv92M/

**Comment by marosset** [comment]
> I only have some minor comments on my first pass here but should probably take a closer look at the shell bits later.
> I see now:
> 
> > We need wincat for a few kubectl proxy scenarios.
> 
> Can we explain that a bit more and check it in somewhere?

@BenTheElder wincat source is checked in at https://github.com/kubernetes-sigs/sig-windows-tools/tree/master/cmd/wincat

**Comment by BenTheElder** [comment]
Is it used for port-forwarding in containerd as well?

In either case, I think we should add a comment in the soruce or a README.md about this, if the runtimes no longer need it for port forwarding (such as with socat on linux) we might want to remove it in the future. As it stands someone will have to go spelunking to find this PR's descriptoin.

File: build/pause/Makefile
Comment (BenTheElder) [reply build/pause/Makefile:97]: thanks!

File: build/pause/windows/pause.c
Comment (BenTheElder) [reply build/pause/windows/pause.c:1]: Yes, but in that case pause.c was the only source code for the image. Now that we have 3 sources in 2 languages, I think we should document them at least a little. As I said it could be a follow up, but I want to mention it here.

**Comment by BenTheElder** [comment]
/approve
/hold
(in case for handling README now)

**Comment by claudiubelu** [comment]
> 
> 
> Is it used for port-forwarding in containerd as well?
> 
> In either case, I think we should add a comment in the soruce or a README.md about this, if the runtimes no longer need it for port forwarding (such as with socat on linux) we might want to remove it in the future. As it stands someone will have to go spelunking to find this PR's descriptoin.

That's a question I've raised myself yesterday too. I've tried it on a containerd env, and no. :) And that's something I'll be working on next.

In any case, I will update the readme shortly.

File: build/pause/Makefile
Comment (claudiubelu) [inline build/pause/Makefile:53]: I've added a comment here mentioning the purpose of the wincat binary. I've also linked this PR, in case anyone needs to read more information about it.

File: build/pause/windows/wincat/wincat.go
Comment (claudiubelu) [inline build/pause/windows/wincat/wincat.go:18]: This also explains the purpose of wincat.

**Review by claudiubelu — COMMENTED** [review]
I've added some comments that explain the purpose of wincat, and I've linked this PR too, in case more information is needed.

Additionally, I'm going to link here the original issues that lead to wincat and some of its bug fixes:

https://github.com/kubernetes/kubernetes/issues/78857
https://github.com/kubernetes/kubernetes/pull/75479
https://github.com/kubernetes-sigs/sig-windows-tools/issues/5

**Comment by claudiubelu** [comment]
Update on the containerd port-forwarding: I was using containerd 1.3.4. I have now updated to 1.4.0, and port-forwarding works for Windows containerd too.

I also found this: https://github.com/containerd/cri/pull/1284

As you can see, wincat is used by containerd too.

**Comment by mkorbi** [comment]
Hello @claudiubelu
This PR has been updated frequently, but it looks like the merge is blocked for some days, so I'd like to check what's the status. The code freeze is starting 12.11.2020 (about 3-4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
As the PR is tagged for 1.20, is it still planned for this release?

**Comment by claudiubelu** [comment]
> 
> 
> Hello @claudiubelu
> This PR has been updated frequently, but it looks like the merge is blocked for some days, so I'd like to check what's the status. The code freeze is starting 12.11.2020 (about 3-4 weeks from now) and while there is still plenty of time, we want to ensure that each PR has a chance to be merged on time.
> As the PR is tagged for 1.20, is it still planned for this release?

This PR has already been approved and LGTM'd, it only has a hold on it. IMO, we can proceed with it, I have answered Ben's question.

**Comment by pacoxu** [comment]
A quick question here: why not build pause image that combined windows and Linux pause image?
Current Linux use pause 3.2 or 3.3 and Windows use 3.4?

**Comment by claudiubelu** [comment]
> .gcr.io/pause:3.4

Hello.

The ``pause:3.4`` image has been built, but it was not published / promoted yet: https://github.com/kubernetes/k8s.io/blob/08581c1cd0937d893f0d1ffdfec8ceb8d7aa8356/k8s.gcr.io/images/k8s-staging-kubernetes/images.yaml

Will send a PR to promote the image.

**Pull request merged**
